### PR TITLE
CASSANDRA-15429: Support NodeTool in dtest

### PR DESCRIPTION
**Pull request by yifan-c**

### CODE REVIEW

File: src/java/org/apache/cassandra/tools/NodeTool.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/tools/NodeTool.java]: Rather than rely on mutable state, can we just have the CLI and core implementation decoupled?  CLI can be simply a ```new NodeProbeFactory.execute(args)``` and the in-jvm dtests use a different implementation?

Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/NodeTool.java]: or make execute a instance method and provide this in the `NodeTool` constructor, so main is ```new System.exit(NodeTool().execute(args))``` and dtest is `new NodeTool(override).execute(args)`

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/tools/NodeProbe.java]: can we use a better name, something like `getMessagingService` or `getMessagingServiceProxy`?

File: test/distributed/org/apache/cassandra/distributed/test/NodeToolTest.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/NodeToolTest.java]: why upgrade?

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: I see your comment, but even then your implementation could still delegate

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: can we refactor `NodeProbe` so this can be easier to maintain?  By adding `getCacheServiceMBean()` there, you could remove this override and just provide that by just returning the field cacheService.

My main concern is that this will become hard to maintain.  As new nodeool commands get added, or existing ones tweaked, we have to know if we should rewrite it here.

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (yifan-c) [reply src/java/org/apache/cassandra/tools/NodeProbe.java]: The method name, `getMsProxy`, is not good. The reason to do so is to follow the existing method naming convention in the class.

File: test/distributed/org/apache/cassandra/distributed/test/NodeToolTest.java
Comment (yifan-c) [reply test/distributed/org/apache/cassandra/distributed/test/NodeToolTest.java]: I will change it to `DistributedTestBase`.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (yifan-c) [reply test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: The `NodeProbe` is messy already. No test cases, new methods added in the class only for the purpose of the new commands (, where the methods should actually belong to), etc.
This is the exact reason I created a mock version and avoid touch the one in prod source. 

In principle, refactoring the `NodeProbe` is welcoming. But I think it deserves a separate PR. 
Adding the cache service field should be part of that...

File: src/java/org/apache/cassandra/tools/NodeTool.java
Comment (yifan-c) [reply src/java/org/apache/cassandra/tools/NodeTool.java]: The `NodeToolCmd` depends on `NodeProbe`. In order to inject the mock dependency, we either passing the mock instance to the cmd instances or like what the patch current have, depends on a static field. Why the `nodeProbeFactory` cannot be an instance field? Because the `NodeToolCmd` is a static inner class, it cannot see unless the field is a class field. 

If passing the mock instance reference, we will end up having the following code in the `NodeTool#execute` method, which I think is ugly. The reason of injecting dependency by doing the type check and casting is that cli parses command into runnable. 

```
            Runnable parse = parser.parse(args);
            if (parse instanceof NodeToolCmd) // hmmm....
            {
                ((NodeToolCmd) parse).setNodeProbeFactory(nodeProbeFactory); // or pass in the probe reference
            }
            printHistory(args);
            parse.run();
```

Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/NodeTool.java]: with a small change, you can get this without statics.

by replacing our usage of `Runnable` with `Consumer<NodeProbeFactory>` (need to extend `Help` to do this), we can replace `parse.run()` with `parse.accept(nodeProbeFactory)`.  Since the usage is just the `connect` method, this won't change the different commands, and will be local to `NodeToolCmd`

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/tools/NodeProbe.java]: there are other naming conversion as well... I guess this class isn't too consistent =)

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: The main issues with this are that it is easier to not be maintained (new `NodeProbe` command added but not updated internal, bugs in existing commands get fixed but not ported to internal, etc.), and by forking you add a larger surface area which could add bugs.  Its possible that the forks change the behavior, and by doing so cause w/e tests use them to pass, but when we use the command on a real cluster it fails.

> The NodeProbe is messy already. No test cases

I 100% agree this is an issue, but by trying to "mock", you actually make it so no tests which use this are actually testing the real code; see point above.

> new methods added in the class only for the purpose of the new commands

And that can keep happening.  Since it can keep happening, the maintenance becomes high if we go with "mock"

> In principle, refactoring the NodeProbe is welcoming. But I think it deserves a separate PR.
Adding the cache service field should be part of that...

My personal preference would be to block this PR on that change.  If you want to do that first, that's fine by me, but I am concerned with the maintenance cost by not doing it, so would rather not move the "mock" in.  

In looking at your changes, a few small refactors would actually be enough to get all your overrides to go away.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: can we refactor `NodeProbe` so `ssProxy` is a function call?  That way this method doesn't need to be overridden

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: can we refactor `NodeProbe` so `ssProxy` is a function call?  That way this method doesn't need to be overridden

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: remove comment

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:103]: i'm ok with this for now, but its possible to implement this by using `CacheService`

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:109]: FYI

```
return StreamSupport.stream(Keyspace.all().spliterator(), false)
                     .flatMap(ks -> ks.getColumnFamilyStores().stream()
                                      .<Map.Entry<String, ColumnFamilyStoreMBean>>map(store -> ImmutablePair.of(ks.getName(), store)))
                     .iterator();
```

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:127]: you could implement this, but it seems you need to handle 3 distinct cases, so fine if this is deferred.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:139]: CompactionManager can return its metrics; see `org.apache.cassandra.db.compaction.CompactionManager#getMetrics`

Annoying with all the special casing, so fine to not implement for now.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:145]: FYI `org.apache.cassandra.metrics.StorageMetrics`. Though the special casing would be annoying.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbeFactory.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbeFactory.java:32]: personal preference (aka your call)

```
return new InternalNodeProbe();
```

Rather than have the dumby constructors in `InternalNodeProbe`

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (yifan-c) [reply test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:109]: Since it is only used for output stats, I will keep it as unsupported. But good to know how to make a columnFamilyStoreIterator.

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/tools/NodeProbe.java:1315]: looking at this again, you could remove this since the field is set properly

Comment (yifan-c) [reply src/java/org/apache/cassandra/tools/NodeProbe.java:1315]: Not sure if I got it. The method is used by another command.

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (dineshjoshi) [inline src/java/org/apache/cassandra/tools/NodeProbe.java]: Nit: the body should appear on the next line.

File: src/java/org/apache/cassandra/tools/NodeProbe.java
Comment (dineshjoshi) [inline src/java/org/apache/cassandra/tools/NodeProbe.java]: Nit: the body should appear on the next line.

File: src/java/org/apache/cassandra/tools/NodeProbeFactory.java
Comment (dineshjoshi) [inline src/java/org/apache/cassandra/tools/NodeProbeFactory.java]: Nit: add new line.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dineshjoshi) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: Please remove unused imports. Following unused imports exist:

```
import java.io.PrintStream;
import java.util.concurrent.TimeUnit;
import javax.management.MalformedObjectNameException;
import org.apache.cassandra.db.ColumnFamilyStore;
import org.apache.cassandra.metrics.DefaultNameFactory;
import org.apache.cassandra.metrics.MetricNameFactory;
import org.apache.cassandra.tools.RepairRunner;
```

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dineshjoshi) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:51]: Is this acceptable? I think we should always have a valid host & port for a `NodeProbe`.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (dineshjoshi) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java]: Don't think any method is going to throw an `IOException`. We can get rid of this.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbeFactory.java
Comment (dineshjoshi) [inline test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbeFactory.java]: Why use inheritance? I would suggest creating a factory interface and have two implementations (Mock and Real). You're not really sharing anything anyway.

**Review by dineshjoshi — CHANGES_REQUESTED** [review]
Once we've completed the review here, I'll review the other branches.

File: test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java
Comment (yifan-c) [reply test/distributed/org/apache/cassandra/distributed/mock/nodetool/InternalNodeProbe.java:51]: Yes. It is intentional. There is no mbean server and the nodetool does not connect to any.

**Pull request closed without merging**
